src/evm/protocol/vm/adapter_contract.rs

```diff
@@ -47,197 +55,144 @@ where
+        })?;
+
+        let price = self.calculate_price(decoded)?;
         Ok(price)
     }
 
     #[allow(clippy::too_many_arguments)]
     pub fn swap(
         &self,
-        pair_id: Vec<u8>,
+        pair_id: String,
         sell_token: Address,
         buy_token: Address,
         is_buy: bool,
         amount: U256,
         block: u64,
         overwrites: Option<HashMap<Address, HashMap<U256, U256>>>,
     ) -> Result<(Trade, HashMap<revm::precompile::Address, StateUpdate>), SimulationError> {
-        let args = vec![
-            Token::FixedBytes(pair_id),
-            Token::Address(H160(sell_token.0 .0)),
-            Token::Address(H160(buy_token.0 .0)),
-            Token::Bool(is_buy),
-            Token::Uint(amount),
-        ];
-
-        let res = self.call("swap", args, block, None, overwrites, None, U256::zero())?;
-
-        let (received_amount, gas_used, price) =
-            if let Token::Tuple(ref return_value) = res.return_value[0] {
-                match &return_value[..] {
-                    [Token::Uint(amount), Token::Uint(gas), Token::Tuple(price_elements)] => {
-                        let received_amount = *amount;
-                        let gas_used = *gas;
-
-                        let price_token = Token::Array(vec![Token::Tuple(price_elements.clone())]);
-                        let price =
-                            self.calculate_price(price_token)?
-                                .first()
-                                .cloned()
-                                .ok_or_else(|| {
-                                    SimulationError::FatalError(
-                                "Adapter swap call failed: An empty price list was returned".into(),
-                            )
-                                })?;
-
-                        Ok((received_amount, gas_used, price))
-                    }
-                    _ => Err(SimulationError::FatalError(format!(
-                        "Adapter swap call failed: Expected amount, gas, and price elements in \
-                    the format (Uint, Uint, Tuple). Found {:?}",
-                        &return_value[..]
-                    ))),
-                }
-            } else {
-                Err(SimulationError::FatalError(format!(
-                "Adapter swap call failed: Expected return_value to be a Token::Tuple. Found {}",
-                res.return_value[0]
-            )))
-            }?;
+        let args = (string_to_bytes32(&pair_id)?, sell_token, buy_token, is_buy, amount);
+        let selector = "swap(bytes32,address,address,uint8,uint256)";
+
+        let res = self.call(selector, args, block, None, overwrites, None, U256::from(0u64))?;
+
+        let decoded: SwapReturn = SwapReturn::abi_decode(&res.return_value, true).map_err(|e| {
+            SimulationError::FatalError(format!("Failed to decode swap return value: {:?}", e))
```

Comment:
Same here - why are we dropping the error details?

Reply:
you right! will add it back!